src/chrome/internal/locations/location.ts

```diff
@@ -0,0 +1,157 @@
+import * as Validation from '../../../validation';
+import { IScript } from '../scripts/script';
+import { ISource } from '../sources/source';
+import { ILoadedSource } from '../sources/loadedSource';
+import { logger } from 'vscode-debugadapter';
+import { ColumnNumber, LineNumber, URLRegexp } from './subtypes';
+import { CDTPScriptUrl } from '../sources/resourceIdentifierSubtypes';
+import { IResourceIdentifier, parseResourceIdentifier, URL } from '../sources/resourceIdentifier';
+
+export type integer = number;
+
+export class Coordinates {
+    public isSameAs(location: Coordinates): boolean {
+        return this.lineNumber === location.lineNumber
+            && this.columnNumber === location.columnNumber;
+    }
+
+    public toString(): string {
+        return this.columnNumber !== undefined
+            ? `${this.lineNumber}:${this.columnNumber}`
+            : `${this.lineNumber}`;
+    }
+
+    constructor(
+        public readonly lineNumber: LineNumber,
+        public readonly columnNumber?: ColumnNumber) {
+        Validation.zeroOrPositive('Line number', lineNumber);
+        if (columnNumber !== undefined) {
+            Validation.zeroOrPositive('Column number', columnNumber);
+        }
+    }
+}
+
+interface ILocation<T extends ScriptOrSourceOrURLOrURLRegexp> {
+    readonly lineNumber: integer;
+    readonly columnNumber?: integer;
+    readonly coordinates: Coordinates;
+    readonly resource: T;
+}
+
+export type ScriptOrSourceOrURLOrURLRegexp = IScript | ILoadedSource | ISource | URLRegexp | URL<CDTPScriptUrl>;
+
+export type Location<T extends ScriptOrSourceOrURLOrURLRegexp> =
+    T extends ISource ? LocationInSource : // Used when receiving locations from the client
+    T extends ILoadedSource ? LocationInLoadedSource : // Used to translate between locations on the client and the debugee
+    T extends IScript ? LocationInScript : // Used when receiving locations from the debugee
+    T extends URLRegexp ? LocationInUrlRegexp : // Used when setting a breakpoint by URL in a local file path in windows, to make it case insensitive
+    T extends URL<CDTPScriptUrl> ? LocationInUrl : // Used when setting a breakpoint by URL for case-insensitive URLs
+    never;
+
+abstract class LocationCommonLogic<T extends ScriptOrSourceOrURLOrURLRegexp> implements ILocation<T> {
+    public get lineNumber(): LineNumber {
+        return this.coordinates.lineNumber;
+    }
+
+    public get columnNumber(): ColumnNumber {
+        return this.coordinates.columnNumber;
+    }
+
+    public toString(): string {
+        return `${this.resource}:${this.coordinates}`;
+    }
+
+    constructor(
+        public readonly resource: T,
+        public readonly coordinates: Coordinates) { }
+}
+
+export class LocationInSource extends LocationCommonLogic<ISource> implements ILocation<ISource> {
```

Comment:
`implements` should be unnecessary on all of these because LocationCommonLogic implements it

Reply:
Removed